Mathlib/Analysis/Convex/ConvexFunction.lean

```diff
@@ -0,0 +1,541 @@
+    rw [e1, e2]; simp; ring_nf; rw [add_comm, ← add_assoc]
+    simp at eq2; rw [← eq2]; nth_rw 1 [← mul_one (f (a • x + b • y))]; rw [← absum1]; ring_nf
+  rw [← inner_smul_right, ← inner_smul_right] at eq3
+  have this1 : b • (y - z) = a • b • (y - x) := by
+    simp; nth_rw 1 [← one_smul ℝ y]; rw [← absum1, add_smul]; simp; rw [← smul_sub, smul_comm]
+  have this2 : a • (z - x) = a • b • (y - x) := by
+    simp; nth_rw 2 [← one_smul ℝ x]; rw [← absum1, add_smul]; simp; rw [← smul_sub, smul_comm]
+  rw [this1, this2, ← inner_sub_left, inner_smul_right, inner_smul_right, ← mul_assoc] at eq3
+  have eq0 : inner (f' (z + d • (y - z)) - f' (x + c • (z - x))) (y - x) = (0 : ℝ) := by
+    contrapose! eq3
+    rw [mul_ne_zero_iff]
+    constructor
+    · rw [mul_ne_zero_iff]; constructor <;> linarith
+    · exact eq3
+  have zeq : z = x + b • (y - x) := by
+    nth_rw 1 [← one_smul ℝ x]; rw [← absum1, add_smul, smul_sub]; simp
+  let u : E := z + d • (y - z)
+  let v : E := x + c • (z - x)
+  have ueq : u = x + (b + d) • (y - x) - d • b • (y - x) := by
+    show z + d • (y - z) = x + (b + d) • (y - x) - d • b • (y - x); rw [zeq]
+    rw [add_assoc, ← add_sub, add_left_cancel_iff]
+    rw [add_smul, ← add_sub, add_left_cancel_iff]
+    rw [← sub_sub, smul_sub]
+  have veq : v = x + c • b • (y - x) := by
+    show x + c • (z - x) = x + c • b • (y - x)
+    rw [zeq, add_left_cancel_iff]; simp
+  have usubv : u - v = (b + d - d * b - c * b) • (y - x) := by
+    rw [ueq, veq, ← smul_assoc, ← smul_assoc, ← sub_sub]; simp
+    rw [← add_sub, ← sub_smul (b + d) (d * b)]; simp; rw [← sub_smul]
+  have eeq0 : inner (f' u - f' v) (u - v) = (0 : ℝ) := by
+    show inner (f' (z + d • (y - z)) - f' (x + c • (z - x))) (u - v) = (0 : ℝ)
+    rw [usubv, inner_smul_right, eq0]; simp
+  have coefne0 : b + d - d * b - c * b > 0 := by
+    nth_rw 1 [← mul_one d]; rw [← absum1]; simp; ring_nf
+    simp at cin; simp at din
+    rcases cin with ⟨_, cl1⟩; rcases din with ⟨dpos, _⟩
+    calc
+      c * b < b := by rw [mul_lt_iff_lt_one_left]; apply cl1; linarith
+      _ < b + d * a := by
+        have : 0 < d * a := by apply mul_pos dpos apos
+        linarith
+  have neq0 : inner (f' u - f' v) (u - v) > (0 : ℝ) := by
+    have uin : u ∈ s := by
+      show z + d • (y - z) ∈ s
+      apply Convex.add_smul_sub_mem hs zin yin; simp; simp at din
+      rcases din with ⟨dpos, dl1⟩; constructor <;> linarith
+    have vin : v ∈ s := by
+      show x + c • (z - x) ∈ s
+      apply Convex.add_smul_sub_mem hs xin zin; simp; simp at cin
+      rcases cin with ⟨cpos, cl1⟩; constructor <;> linarith
+    apply mono u uin v vin
+    by_contra h
+    have : u - v = 0 := by rw [h]; simp
+    rw [usubv, smul_eq_zero] at this;
+    contrapose! this
+    constructor
+    · linarith
+    · rw [sub_ne_zero]; symm; exact xney
+  linarith
+
+theorem strict_convex_monotone_gradient (hf : ∀ x ∈ s, HasGradientAt f (f' x) x)
+    (h₁ : StrictConvexOn ℝ s f ) :
```

Comment:
There are quite a few of these extra spaces throughout the file that you should remove eventually.
```suggestion
    (h₁ : StrictConvexOn ℝ s f) :
```